tests/batch_size_optimizer/test_simulator.py

```diff
@@ -0,0 +1,89 @@
+from fastapi.testclient import TestClient
+from pyparsing import Literal
+import pytest
+from pytest_mock import MockerFixture
+from examples.trace_driven.run_single import read_trace
+from zeus.job import Job
+from zeus.optimizer.batch_size.server.router import app
+from zeus.policy.optimizer import JITPowerLimitOptimizer, PruningGTSBatchSizeOptimizer
```

Comment:
Hmm, so we wanted to make the BSO production-ready by making it an actual server, which will be under `zeus.optimizer.batch_size.server`. However, for the sake of the research community which will try to compare with Zeus, it's still better to leave the simulator in a functioning state.

Guessing from the fact that you imported from `zeus.policy` in tests, I'm guessing you are going to leave `zeus.policy` as is? Then what is `zeus/optimizer/batch_size/server/mab.py` and stuff? Are they basically duplicated code? Is duplication necessary? Ultimately, what would be the best way to support both the server and the simulator?

Reply:
Or perhaps the `pytest` simulator suite you're building (porting from `examples/trace_driven`) can basically be used as the simulator?

Reply:
The only reason I imported policy is to compare the simulator's result with the old one so that I can check new optimizer is giving the same output. I can just hardcode the selected batch_sizes and remove policy.optimizer.

Reply:
I'm fine with moving stuff in `zeus.policy` into `zeus.optimizer.batch_size`. I'm actually not entirely sure about hardcoding selected batch sizes. Say one day someone makes a reasonable change to the MAB, and the test fails. Then it's kinda hard to debug. Rather the tests are better off checking consistency between the simulator and the BSO server. If the simulator and BSO server basically share the **exact** same logic but just wrap it differently, only a couple obvious test cases derived from `examples/trace_driven` would be enough.

I'm not really getting the big picture I guess. Could you briefly write up how the simulator will look after this PR?

Reply:
My initial idea about the simulator was basically using a new BSO.

1. Use a new BSO (Dummy BSO client that sends requests to the server).
2. Use the BSO server using `fastapi.testclient` so that it doesn't have to run the server. The test client will run the server and respond to the dummy BSO client's request for simulation.

Now I think this might get complicated after we integrate DB since we have to mock DB too.

Reply:
I see. It's basically "mocking the server including the DB" vs. "Rewriting the simulator using core logic you migrated from `zeus.policy` into `zeus.optimizer.batch_size`". The former approach will force us to probably spin up sqlite to store state... which I agree it's too much overhead for a dumb simulator.

I have a pretty radical idea. The simulator is not Zeus's main offering. It's for the convenience of the research community. What about just entirely duplicating all code and moving all the simulator related stuff into `zeus._legacy`? Instead of moving stuff from `zeus.policy` into `zeus.optimizer.batch_size`, you *copy*. Then, all the parts that are now useless from the perspective of `zeus.optimizer.batch_size` moves into `zeus._legacy` and we just update `examples/trace_driven` slightly so that it works again. WDYT?

Reply:
I see. I think that should make things clean. 
For "the BSO server should be tested as if the simulator does not exist," are we still doing a trace_driven simulation (just like we are doing now using `simulate_with_server.py`?
Also, when you said "update `examples/trace_driven`", do we assume that the user will run the server, so no mocking for DB?

Reply:
> For "the BSO server should be tested as if the simulator does not exist," are we still doing a trace_driven simulation (just like we are doing now using `simulate_with_server.py`?

Basically the core brain of the BSO is the MAB. Perhaps the pruning batch size explorer. Pretty much all other things are just wrappers around them. I think now is the time where we port simulation results from `examples/trace_driven` into **MAB test cases**. That is, if you run the simulator, it runs an instance of an MAB inside to select the batch sizes and prints out arm states. Can you port five different cases to create five MAB test cases? You can directly hardcode data, including arm selection order.

>Also, when you said "update `examples/trace_driven`", do we assume that the user will run the server, so no mocking for DB?

No I think it'll be something like `sed -i 's/zeus/zeus._legacy/g'`.